src/common/transformations/src/transformations/symbolic_transformations/symbol_optimization.cpp

```diff
@@ -218,25 +219,106 @@ void optimize_value_usage(ov::Output<ov::Node>& output, STS_map& symbol_shape_so
+                for (const auto& input : output.get_target_inputs()) {
+                    const auto& output_op = input.get_node();
+                    const auto& rt_info = output_op->get_rt_info();
+                    if (!rt_info.count(weight_rt_info_name))
+                        continue;
+                    output_has_weight = true;
+                    auto output_weight = rt_info.at(weight_rt_info_name).as<int64_t>();
+                    weight = output_weight > weight ? output_weight : weight;
+                }
+            }
+            if (output_has_weight) {
+                if (op->get_rt_info()[op_depends_on_parameter].as<bool>()) {
+                    weight += 1;
+                }
+                op->get_rt_info()[weight_rt_info_name] = weight;
+            }
+        }
+    }
+    // step 3: make propagation for all the nodes:
+    // if weight is already assigned -- skip operation
+    // else operation weights is minimum of input indices
+    // if all operation inputs have no weights -- this op is isolated and this algorithm doesn't make sense,
+    // such cases are extremely rare and rather theoretical, to handle them we return original ov::Model op order
+    std::map<int64_t, std::vector<std::shared_ptr<ov::Node>>> level_to_vector;
+    for (const auto& op : m->get_ordered_ops()) {
+        if (op->get_rt_info().count(weight_rt_info_name)) {
+            level_to_vector[op->get_rt_info().at(weight_rt_info_name).as<int64_t>()].push_back(op);
             continue;
-        symbol_shape_source[symbol] = output;
+        } else {
+            int64_t weight = 0;
+            bool inputs_have_weight = false;
+            for (const auto& input : op->input_values()) {
+                const auto& rt_info = input.get_node_shared_ptr()->get_rt_info();
+                if (!rt_info.count(weight_rt_info_name))
+                    continue;
+                auto input_weight = rt_info.at(weight_rt_info_name).as<int64_t>();
+                if (!inputs_have_weight)
+                    weight = input_weight;
+                weight = input_weight < weight ? input_weight : weight;
+                inputs_have_weight = true;
+            }
+            if (inputs_have_weight)
+                op->get_rt_info()[weight_rt_info_name] = weight;
+            else
+                return m->get_ordered_ops();
+            level_to_vector[weight].push_back(op);
+        }
     }
+    // finalization: descending order for levels and ops within level are ordered by get_ordered_ops
+    std::vector<std::shared_ptr<ov::Node>> result;
+    result.reserve(reverse_order.size());
+    for (const auto& item : level_to_vector)
+        result.insert(result.end(), item.second.rbegin(), item.second.rend());
+    for (const auto& op : result) {
+        op->get_rt_info().erase(weight_rt_info_name);
+        op->get_rt_info().erase(op_depends_on_parameter);
+    }
+    std::reverse(result.begin(), result.end());
```

Comment:
Is possible to insert into `result` in reverse order then using reverse would not be required?